### PR TITLE
Onnx backend

### PR DESCRIPTION
**Pull request by clsandoval**
## Description

This PR adds a complete ONNX backend for PyTensor, enabling export of PyTensor computational graphs to ONNX format and execution via ONNX Runtime.

**Key features:**
- **ONNX Linker**: Core infrastructure for converting PyTensor graphs to ONNX models
- **High-level export API**: `export_onnx()`, `compile_onnx()`, and `export_function_onnx()` for easy model export
- **Comprehensive dispatcher system**: Type-based dispatch for converting PyTensor Ops to ONNX nodes
- **Operation support across multiple categories**:
  - Elementwise operations (Add, Sub, Mul, Div, Exp, Log, trigonometric functions, etc.)
  - Shape operations (Reshape, DimShuffle, Transpose, Join, Split)
  - Subtensor operations (basic slicing, advanced indexing, IncSubtensor)
  - Math operations (Sum, Prod, Max, Min, Mean, Argmax, Argmin)
  - Linear algebra (MatMul, Dot, BatchedDot, matrix operations)
  - Neural network ops (Softmax, LogSoftmax, Sigmoid, Conv2d, MaxPool)
- **Property-based testing**: Hypothesis-powered test strategies for robust validation

**Files added:**
- `pytensor/link/onnx/` - Core ONNX backend implementation (13 files, ~2k lines)
- `tests/link/onnx/` - Comprehensive test suite (18 files, ~4k lines, 186 tests)

## Related Issue

- [ ] Closes #
- [ ] Related #

## Checklist

- [x] Checked that [the pre-commit linting/style checks pass](https://docs.pymc.io/en/latest/contributing/python_style.html)
- [x] Included tests that prove the fix is effective or that the new feature works
- [ ] Added necessary documentation (docstrings and/or example notebooks)
- [ ] If you are a pro: each commit corresponds to a [relevant logical change](https://wiki.openstack.org/wiki/GitCommitMessages#Structural_split_of_changes)

## Type of change

- [x] New feature / enhancement
- [ ] Bug fix
- [ ] Documentation
- [ ] Maintenance
- [ ] Other (please specify):

### CODE REVIEW

File: pytensor/link/onnx/dispatch/basic.py [hunk 1]
```diff
@@ -0,0 +1,324 @@
...
+        Data type
+    name : str, optional
+        Name for the tensor
+
+    Returns
+    -------
+    onnx.TensorProto
+        ONNX tensor representation
+    """
+    # Default: try to convert to numpy array first
```
Comment (ricardoV94) [inline pytensor/link/onnx/dispatch/basic.py:46]: A better default is to raise, we did this before for other backends and have been moving away

File: pytensor/link/onnx/dispatch/basic.py [hunk 2]
```diff
@@ -0,0 +1,324 @@
...
+        raise ValueError(
+            f"Unsupported dtype: {dtype_str}. "
+            f"Supported dtypes: {list(PYTENSOR_DTYPE_TO_ONNX.keys())}"
+        )
+    onnx_dtype = PYTENSOR_DTYPE_TO_ONNX[dtype_str]
+
+    # Get shape - handle both static and symbolic shapes
+    # For now, we'll use None for unknown dimensions
+    ndim = var.type.ndim
+    shape = [None] * ndim  # Unknown dimensions
```
Comment (ricardoV94) [inline pytensor/link/onnx/dispatch/basic.py:120]:
```suggestion
    shape = var.type.shape
```

Comment (ricardoV94) [reply pytensor/link/onnx/dispatch/basic.py:120]: What about non-TensorVariables? Are we raising explicitly if not supported? Examples include Slices, TypedLists, RandomGenerator, SparseTensorVariables

File: pytensor/link/onnx/dispatch/basic.py [hunk 3]
```diff
@@ -0,0 +1,324 @@
...
+    Returns
+    -------
+    onnx.ModelProto
+        Complete ONNX model
+    """
+    # Track variable names to ensure uniqueness
+    var_names = {}
+    var_counter = 0
+
+    def get_var_name(var):
```
Comment (ricardoV94) [inline pytensor/link/onnx/dispatch/basic.py:184]: there's a unique_name_generator helper already in link.utils that I think you can reuse

File: pytensor/link/onnx/dispatch/basic.py [hunk 4]
```diff
@@ -0,0 +1,324 @@
...
+        if isinstance(var, Constant):
+            name = get_var_name(var)
+            # Convert constant to ONNX initializer
+            # Special handling: if constant is a scalar int type and is used in operations
+            # with float tensors, upcast to float32 to avoid type mismatches
+            data = var.data
+            if data.ndim == 0 and np.issubdtype(data.dtype, np.integer):
+                # Check if this constant is used with float operations
+                # For now, we'll upcast all scalar integer constants to float32
+                # This is a simplification but handles the common case of: x * 2
```
Comment (ricardoV94) [inline pytensor/link/onnx/dispatch/basic.py:213]: This doesn't sound safe. Constants show up a lot in indexing operations for example x[:2], you wouldn't want to make that a float. Any implicit casting should be done by the Op that needs it, or is there a more fundamental onnx limitation here?

File: pytensor/link/onnx/dispatch/basic.py [hunk 5]
```diff
@@ -0,0 +1,324 @@
...
+                if onnx_node is not None:
+                    nodes.append(onnx_node)
+                if node_initializers:
+                    initializers.extend(node_initializers)
+            else:
+                # Returned single node (most common case)
+                # Example: Add returns single Add node
+                nodes.append(result)
+        else:
+            # Handler returned None - this is a no-op operation
```
Comment (ricardoV94) [inline pytensor/link/onnx/dispatch/basic.py:252]: If you really need this: Make the handler return a specific sentinel `ONNX_NO_OP` instead of `None` to avoid subtle errors where users just forget to return something?

Given you have identity node, below it sounds like you don't need it though

File: pytensor/link/onnx/dispatch/basic.py [hunk 6]
```diff
@@ -0,0 +1,324 @@
...
+        ir_version=9,  # Use IR version 9 for ONNX Runtime compatibility
+    )
+
+    # Check the model
+    onnx.checker.check_model(model_def)
+
+    return model_def
+
+
+@onnx_funcify.register(Constant)
```
Comment (ricardoV94) [inline pytensor/link/onnx/dispatch/basic.py:302]: Constants aren't nodes so you shouldn't need to register them for funcify

File: pytensor/link/onnx/dispatch/basic.py [hunk 7]
```diff
@@ -0,0 +1,324 @@
...
+def onnx_funcify_DeepCopyOp(op, node, get_var_name, **kwargs):
+    """Convert DeepCopyOp to ONNX Identity node.
+
+    DeepCopyOp is equivalent to Identity in ONNX.
+    """
+    input_names = [get_var_name(inp) for inp in node.inputs]
+    output_names = [get_var_name(out) for out in node.outputs]
+
+    return helper.make_node(
+        "Identity",
```
Comment (ricardoV94) [inline pytensor/link/onnx/dispatch/basic.py:320]: If you have Identity in ONNX why can't you use those for the Ops like `SpecifyShape` that you are handling with special casing in the FunctionGraph loop?

File: pytensor/link/onnx/dispatch/elemwise.py [hunk 8]
```diff
@@ -0,0 +1,267 @@
...
+
+from onnx import helper
+
+from pytensor.link.onnx.dispatch.basic import onnx_funcify
+from pytensor.scalar import basic as scalar
+from pytensor.scalar import math as scalar_math
+from pytensor.tensor.elemwise import Elemwise
+
+
+# ⭐ THE MAGIC MAPPING - Tier 1 + Tier 4-5 operations
```
Comment (ricardoV94) [inline pytensor/link/onnx/dispatch/elemwise.py:11]: Fun nazi:
```suggestion
```

File: pytensor/link/onnx/dispatch/elemwise.py [hunk 9]
```diff
@@ -0,0 +1,267 @@
...
+        Additional keyword arguments
+
+    Returns
+    -------
+    onnx.NodeProto or list[onnx.NodeProto]
+        ONNX node(s) for the operation
+    """
+    scalar_op_type = type(op.scalar_op)
+
+    # Special handling for operations that need to be composed
```
Comment (ricardoV94) [inline pytensor/link/onnx/dispatch/elemwise.py:93]: Suggestion: after checking an entry is not in the standard map, use another layer of singledispatch, on the core_op so it can be extended from the outside, instead of this closed-box nested if-else for special cases

File: pytensor/link/onnx/dispatch/elemwise.py [hunk 10]
```diff
@@ -0,0 +1,267 @@
...
+    # Clip(x, min, max) - ONNX requires scalar min/max, but PyTensor may provide tensors
+    if scalar_op_type == scalar.Clip:
+        input_names = [get_var_name(inp) for inp in node.inputs]
+        output_name = get_var_name(node.outputs[0])
+
+        # Input 0 is the array to clip, inputs 1 and 2 are min/max
+        # ONNX Clip expects scalars for min/max, but PyTensor may have added dimensions
```
Comment (ricardoV94) [inline pytensor/link/onnx/dispatch/elemwise.py:100]: This can change the semantics of the graph, if in the PyTensor graph the bounds were broadcasting the input variables. Introspect the node input types to be sure it would be equivalent, and raise `NotImplementedError` otherwise

File: pytensor/link/onnx/dispatch/nlinalg.py [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+        alpha=alpha,
+        beta=beta,
+        transA=0,
+        transB=0,
+    )
+
+    return gemm_node
+
+
+@onnx_funcify.register(BatchedDot)
```
Comment (ricardoV94) [inline pytensor/link/onnx/dispatch/nlinalg.py:80]: can't you register to the same function you registered `Dot`?

File: pytensor/link/onnx/dispatch/nlinalg.py [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+        "MatMul",
+        inputs=[input_a, input_b],
+        outputs=[output_name],
+        name=f"MatMul_{output_name}",
+    )
+
+    return matmul_node
+
+
+@onnx_funcify.register(Gemm)
```
Comment (ricardoV94) [inline pytensor/link/onnx/dispatch/nlinalg.py:34]: If onnx includes the BlasOpt rewrites that introduce Gemm, it will also introduce stuff like Dot22, Dot22Scalar, GemV, Ger, so make sure to either handle all, or not include the rewrites when linking to onnx

Comment (ricardoV94) [reply pytensor/link/onnx/dispatch/nlinalg.py:34]: Also rename this file to linalg.py perhaps

File: pytensor/link/onnx/dispatch/nnet.py [hunk 13]
```diff
@@ -0,0 +1,184 @@
+"""ONNX conversion for neural network operations."""
```
Comment (ricardoV94) [inline pytensor/link/onnx/dispatch/nnet.py:1]: Optional: missing the SoftmaxGrad

Suggestion: put this in `special.py`, the concept of nnet is an onnx one, not a pytensor one

File: pytensor/link/onnx/dispatch/shape.py [hunk 14]
```diff
@@ -0,0 +1,382 @@
...
+from onnx import helper, numpy_helper
+
+from pytensor.graph.basic import Constant
+from pytensor.link.onnx.dispatch.basic import onnx_funcify
+from pytensor.tensor.basic import Join, Split, get_scalar_constant_value
+from pytensor.tensor.exceptions import NotScalarConstantError
+from pytensor.tensor.shape import Reshape, Shape, Shape_i, SpecifyShape
+
+
+@onnx_funcify.register(type(None))
```
Comment (ricardoV94) [inline pytensor/link/onnx/dispatch/shape.py:13]: None are not Ops, but data/ variable types

File: pytensor/link/onnx/dispatch/shape.py [hunk 15]
```diff
@@ -0,0 +1,382 @@
...
+
+    # Return list of nodes - this is the key pattern!
+    return [idx_constant, shape_node, gather_node]
+
+
+@onnx_funcify.register(SpecifyShape)
+def onnx_funcify_SpecifyShape(op, node, get_var_name, **kwargs):
+    """SpecifyShape is just a hint - pass through input.
+
+    SpecifyShape doesn't change the tensor data, it just provides
```
Comment (ricardoV94) [inline pytensor/link/onnx/dispatch/shape.py:108]: Not true, SpecifyShape is a runtime Assertion as well. Reword or handle if onnx supports shape assertions without too much overhead

File: pytensor/link/onnx/dispatch/shape.py [hunk 16]
```diff
@@ -0,0 +1,382 @@
...
+        else:
+            raise NotImplementedError(
+                f"DimShuffle with new_order={new_order} and input_ndim={input_ndim} "
+                f"is not yet supported in ONNX backend."
+            )
+
+
+except ImportError:
+    # DimShuffle not available
+    pass
```
Comment (ricardoV94) [inline pytensor/link/onnx/dispatch/shape.py:212]: remove

File: pytensor/link/onnx/dispatch/subtensor.py [hunk 17]
```diff
@@ -0,0 +1,436 @@
...
+        x = pt.vector('x')
+        indices = pt.vector('indices', dtype='int64')
+        y = x[indices]  # AdvancedSubtensor (gets optimized to AdvancedSubtensor1 in normal mode)
+
+        ONNX: Gather(x, indices, axis=0)
+    """
+    # For now, we only handle the simple case that matches AdvancedSubtensor1
+    # More complex cases would need GatherND or multiple operations
+
+    if len(node.inputs) != 2:
```
Comment (ricardoV94) [inline pytensor/link/onnx/dispatch/subtensor.py:219]: probably also need to check indices is a vector or integers (instead of a matrix or a boolean mask). It could also be a basic index, although that would be odd to find as the only index of an AdvancedSubtensor Op, be defensive just in case

File: pytensor/link/onnx/rewrite.py [hunk 18]
```diff
@@ -0,0 +1,48 @@
...
+import numpy as np
+
+from pytensor import scalar as ps
+from pytensor.graph.rewriting.basic import node_rewriter
+from pytensor.tensor.elemwise import Elemwise
+from pytensor.tensor.math import add, exp, log, sub
+
+
+@node_rewriter([Elemwise])
+def expand_log1p_expm1_for_onnx(fgraph, node):
```
Comment (ricardoV94) [inline pytensor/link/onnx/rewrite.py:16]: Didn't you handle this in Elemwise funcify?

File: tests/link/onnx/conftest.py [hunk 19]
```diff
@@ -0,0 +1,57 @@
...
+@pytest.fixture
+def float32_vector(rng):
+    """Sample float32 vector for testing."""
+    return rng.normal(size=10).astype("float32")
+
+
+@pytest.fixture
+def float32_matrix(rng):
+    """Sample float32 matrix for testing."""
+    return rng.normal(size=(5, 5)).astype("float32")
```
Comment (ricardoV94) [inline tests/link/onnx/conftest.py:57]: Personal preference: these fixtures don't help understanding tests and they don't save much code either

File: tests/link/onnx/test_basic.py [hunk 20]
```diff
@@ -0,0 +1,169 @@
...
+def test_get_onnx_node_types():
+    """Test that get_onnx_node_types utility works."""
+    pytest.importorskip("onnx")
+    pytest.importorskip("onnxruntime")
+
+    import pytensor
+    import pytensor.tensor as pt
+    from pytensor.link.onnx.linker import ONNXLinker
+
+    # Create a graph with Add operation
```
Comment (ricardoV94) [inline tests/link/onnx/test_basic.py:158]: If you iterate on this further, please ask it to remove silly comments like this:

```
# do_x
do_x()
```

File: tests/link/onnx/test_elemwise.py [hunk 21]
```diff
@@ -0,0 +1,526 @@
...
+    - Log operation works with positive inputs
+    - ONNX output matches Python reference
+    - Correct ONNX node type (Log) is generated
+
+    Note: Uses positive_float32_array_strategy to ensure valid inputs
+          (log requires x > 0). Uses 50 examples (vs standard 10) due to
+          numerical sensitivity.
+    """
+    pytest.importorskip("onnx")
+    pytest.importorskip("onnxruntime")
```
Comment (ricardoV94) [inline tests/link/onnx/test_elemwise.py:148]: Don't put these in each test

File: tests/link/onnx/test_elemwise.py [hunk 22]
```diff
@@ -0,0 +1,526 @@
...
+    x = pt.vector("x", dtype="float32")
+    y = pt.sqrt(x)
+
+    x_val = np.array([1, 4, 9, 16], dtype="float32")
+
+    fn, _result = compare_onnx_and_py([x], y, [x_val])
+    assert "Sqrt" in get_onnx_node_types(fn)
+
+
+def test_pow():
```
Comment (ricardoV94) [inline tests/link/onnx/test_elemwise.py:457]: Can many of these elemwise tests be combined with a `pytest.mark.parametrize`?

File: tests/link/onnx/test_extra_ops.py [hunk 23]
```diff
@@ -0,0 +1,102 @@
...
+
+    x_val = np.array([1, 2, 3], dtype="float32")
+
+    _fn, result = compare_onnx_and_py([x], y, [x_val])
+
+    expected = np.repeat(x_val, repeats=3, axis=0)
+    np.testing.assert_array_equal(result, expected)
+
+
+# Unique Tests
```
Comment (ricardoV94) [inline tests/link/onnx/test_extra_ops.py:52]: Remove these redundant comments. They are also esay to go out of place when we refactor tests around

File: tests/link/onnx/test_strategies.py [hunk 24]
```diff
@@ -0,0 +1,299 @@
...
+
+import pytensor.tensor as pt
+
+
+# ============================================================================
+# REGISTRY STRUCTURE TESTS
+# ============================================================================
+
+
+def test_elemwise_registry_exists():
```
Comment (ricardoV94) [inline tests/link/onnx/test_strategies.py:20]: Why test this? Sounds like an implementation details of little consequence

File: tests/link/onnx/test_subtensor.py [hunk 25]
```diff
@@ -0,0 +1,482 @@
...
+
+
+# ============================================================================
+# MANUAL EDGE CASE TESTS
+# ============================================================================
+# These tests complement the property-based tests above by:
+# - Testing specific edge cases and patterns
+# - Providing readable examples for documentation
+# - Validating 3D operations (more complex than property tests cover)
+# ============================================================================
```
Comment (ricardoV94) [inline tests/link/onnx/test_subtensor.py:243]: Put comments inside test functions / class so they don't get lost in future refactors. Do err on the side of less comments

File: uv.lock [hunk 26]
```diff
@@ -0,0 +1,1083 @@
+version = 1
```
Comment (ricardoV94) [inline uv.lock:1]: remove uv.lock file. If you want to argue for it, do it in a separate PR

File: tests/link/onnx/strategies.py [hunk 27]
```diff
@@ -0,0 +1,748 @@
+"""Hypothesis strategies and operation registries for ONNX backend testing."""
```
Comment (ricardoV94) [inline tests/link/onnx/strategies.py:1]: How should I review this 700 line file? Do we need it as a monolith?

**Review by ricardoV94 — COMMENTED** [review]
Hi @clsandoval 

Thanks for opening the PR. Looks pretty promising.

I left a bunch of comments. In general this would have been nicer split into a PR that implements the basics, followed by another that adds most functionality.

If you intend to iterate on this PR to address the comments, please make sure to manually validate against LLM amnesia, I don't want to be policing against undoing of previous agreed changes and stuff of that sort in a PR that has 8k LOC changes.

Also please prompt it to avoid comment galore. I can read the function names. Also tell it avoid stuff like `# now testing axis=2` because we told it the case was missing in a previous review.



With that grumpy note behind, Great work!

PS: edit the test yaml config, so onnx tests are only run in a single separate job, and not mixed with the previous tests

File: tests/link/onnx/test_elemwise.py (hunk 21 above)
Comment (ricardoV94) [reply tests/link/onnx/test_elemwise.py:148]: pre-commit can be told to ignore out of order imports

File: pytensor/link/onnx/dispatch/elemwise.py (hunk 8 above)
Comment (jessegrabowski) [reply pytensor/link/onnx/dispatch/elemwise.py:11]: This reminds me of when you told me I couldn't use 69420 as a random seed
